hs_core/models.py

```diff
@@ -1750,70 +1766,494 @@ class Meta:
+    """
+    def __init__(self):
+        super(FedStorage, self).__init__("federated")
+
+
+# TODO: revise path logic for rename_resource_file_in_django for proper path.
+# TODO: utilize antibugging to check that paths are coherent after each operation.
 
 class ResourceFile(models.Model):
+    """
+    Represent a file in a resource.
+    """
+
+    # A ResourceFile is a sub-object of a resource, which can have several types.
     object_id = models.PositiveIntegerField()
     content_type = models.ForeignKey(ContentType)
     content_object = GenericForeignKey('content_type', 'object_id')
-    # This is used to direct uploads to a subfolder of the root folder. See get_path above.
-    file_folder = models.CharField(max_length=4096, null=True, blank=True)
-    resource_file = models.FileField(upload_to=get_path, max_length=500,
-                                     null=True, blank=True,
-                                     storage=IrodsStorage())
-    # the following optional fields are added for use by federated iRODS resources where
-    # resources are created in the local federated zone rather than hydroshare zone, in
-    # which case resource_file is empty, and we record iRODS logical resource file name
-    # and file size for customized copy or move operations for resource creation and other
-    # operations fed_resource_file in particular is a counterpart of resource_file, but
-    # handles files uploaded from local disk and store the files to federated zone rather
-    # than hydroshare zone.
-    fed_resource_file = models.FileField(upload_to=get_path, max_length=500,
-                                         null=True, blank=True,
-                                         storage=IrodsStorage('federated'))
-    fed_resource_file_name_or_path = models.CharField(max_length=255,
-                                                      null=True, blank=True)
-    fed_resource_file_size = models.CharField(max_length=15,
-                                              null=True, blank=True)
+
+    # This is used to direct uploads to a subfolder of the root folder for the resource.
+    # See get_path and get_resource_path above.
+    file_folder = models.CharField(max_length=4096, null=True)
+
+    # This pair of FileFields deals with the fact that there are two kinds of storage
+    resource_file = models.FileField(upload_to=get_path, max_length=4096,
+                                     null=True, blank=True, storage=IrodsStorage())
+    fed_resource_file = models.FileField(upload_to=get_path, max_length=4096,
+                                         null=True, blank=True, storage=FedStorage())
+
+    # DEPRECATED: utilize resfile.set_storage_path(path) and resfile.storage_path.
+    # fed_resource_file_name_or_path = models.CharField(max_length=255, null=True, blank=True)
+    # DEPRECATED: use native size routine
+    # fed_resource_file_size = models.CharField(max_length=15, null=True, blank=True)
```

Comment:
@alvacouch After reviewing a bit more, it looks like you have tried to use fed_resource_file to cover this third scenario I mentioned above. This may be sufficient and definitely simplify things a lot. I will have to test it out to make sure each scenario works as expected, though. It'd be great if we can deploy this to a VM where local iRODS containers are used so we can test federation. @mjstealey Can we make a VM self-deployment to use local iRODS containers so that we can test federations? It'd be the best to test this PR in a self-deployed VM before it is merged to develop.